### ISSUE TITLE
smartPointing data?

### ISSUE THREAD

**Issue by gpasquie** [medInria/medInria-public#226]
We thought having solved all the problems of this kind by smartpointing in the medAbstractDatabaseImporter but actually, there are lots of cases where it can also crash...
For example, just do:
```c++
    medImageMaskAnnotationData *annotation = new medImageMaskAnnotationData;
    // d->mask is an medAbstractImageData containing an itk image
    annotation->setMaskData(d->mask);
    view->addLayer(annotation);
```
and it's going to crash on the addLayer line.
When you add the data to the view, it's going to be smartPointed and stored in a QList.
When you manipulate the view, you access the list and some operations are made on the list (insert, contains, ...). These operations can retain/release the smartPointer. In our case, the annotation is not smartpointed at creation time, so the number of reference is only one, and the release will delete the annotation... :(

Can we prevent anyone to create non smartPointed data?
Should we handle the case where data is not smartPointed?

**Comment by mikejbuckingham** [comment]
So for me, I'd suggest that we enforce smartpointer usage through the method you talked about during the meeting - specifically having a method to handle creation of the object, and keeping the constructor protected.

I don't really see how we can explicitly handle the situation where the data isn't owned by a smartpointer, as view can't know anything about the ownership of the resource (unless we fancy implementing a full memory manager, which I think is probably a bit overkill!)

**Comment by fvichot** [comment]
OK, so this my thoughts on the matter: the original problem was that we had too many leaks, many of which were caused by excessive smartPointing, hence my then suggestion that we should only smartPoint when really necessary. So we removed most smartPointers. But we're now in the situation of inconsistent smartPointing, where we get the completely unexpected and unacceptable behaviour of getting your object deleted by calling ```addLayer``` or ```generateThumbnail``` on it.

So, there's a few options : 
1. Continue with the convention of "smart smartPointing", where you're supposed to know what you're doing, and know which calls could delete your data and thus smartPoint. I think we agree this is not the way to go.
2. Go the VTK route, and force all allocations to use a ```Type::New()``` static method, which returns a raw pointer, but with the ref count incremented. My issue with this is that you need to remember to call ```Object->Delete()```, otherwise, we get leaks. I fear we'll just get back to how it was and get occasional leaks of many hundreds of MBytes because a ```Delete()``` was fogotten.
3. Or, a similar but I think better option: make all constructors private/protected, and force usage of a static ```New()``` method, but have it return a smartPointer everytime. That way, you can't forget to call delete, you can't have it deleted by calls to ```addLayer()```, and having to store it in a smartPointer, it should remind you that this WILL get deleted once it reaches the end of the scope, unless you save it somewhere. I think the destructors should be made protected too, this way no ```delete``` can be called by error on them.
This way, we don't need to modify methods with raw pointers in their prototype, just call them by providing smartPointedObject.data() as the pointer. This means we can call ```this``` as an argument, which should be garanteed to not pose problams, as if there is a ```this```, there is a smartPointer to that object somewhere. The method is then free to manipulate the raw pointer without risks of it being deleted (as there is garantee that another method up the stack has reference to it) or re-wrap it on a smartPointer if it wants to keep a ref to it.

**To summarize:**
- We smartPoint EVERY instantiation of data objects. I suggest to enforce it by making constructors and destructors protected, and adding a New() static method.
- Either passing raw pointers or smart pointers to method is fine.
- Functions systematically return smartPointers if they return data object pointers. It's actually not always necessary, but I think it's one of those rules where it's just easier to always do it than document the few cases we it would be fine not to. This means modifying the data factory as well.
- NEVER store anything other than smartPointers to data object, no raw pointers.

And added bonus, we can switch the dataManager to QPointers and remove my garbage collect hack which I like less and less.

I think this covers everything. Thoughts ?

**Comment by rdebroiz** [comment]
Ok I'm not sure to understand all of it but it sounds good. Just to clarify, if we take the third option, it means that:

- A typical creation would be someting like: `medAbstractData::Pointer data = medAbstractData::new()`
- A typical use of this data as parameter of a method would be: `foo->doSomething(data.rawPointer())` or  `foo->doSomething(this)`  if we are in medAbstractData scope
- Finally our smart pointer are still dtkSmartPointer

Am i wrong ? 

About:
> - Either passing raw pointers or smart pointers to method is fine.
- NEVER store anything other than smartPointers to data object, no raw pointers. 

```c++
class foo
{
private:
   QList<medAbstractData::Pointer> dataList;
public:
   void setData(medAbstractData *data)
   {
        dataList << medAbstractData::Pointer(data);
   }
}
```
Does this snippet illustrates well what you said ?

**Comment by fvichot** [comment]
Yes,Yes,Yes,No,Yes. :wink:
